---
title: "A guide to using cellexalvrR"
author: "Shamit Soneji"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette
  
vignette: >
  %\VignetteIndexEntry{cellexalvrR guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

This document describes how to use `cellexalvrR`, an R package that accompanies [CellexalVR](wwww.cellexalvr.med.lu.se) which is a virtual reality environment to analyze single-cell RNAseq data. `cellexalvrR` has two functions:

1. To aid the formatting and export of data that can be imported by CellexalVR.
2. To perform backend calculations during a CellexalVR session.

##Installation
The easiest way to install `cellexalvrR` is directly from github using the `devtools` package:

```{r,eval=F}
library(devtools)
install_github("sonejilab/cellexalvrR")
```

**If you are installing this on your VR station then be sure to install `cellexalvrR` system-wide, so on your Windows 10 machine right-click the R icon, and then "Run as administrator" before issuing the commands above.**

Make sure that you also have the Rtools installed on your windows machine which provides c and c++ compilers.

## Quick start

The data from [Nestorowa *et al*](http://www.bloodjournal.org/content/128/8/e20.long?sso-checked=true) can be downloaded from [here](https://www.cellexalvr.med.lu.se/download#). Unpack them and set your working directory to where they are.

First, load the library:

```{r,message=FALSE}
library(cellexalvrR)

if ( ! file.exists("log2data.RData") ) {
   log2data = as.matrix(cellexalObj@data)
   save( log2data, file="log2data.RData")
   rm( log2data)
   
   facs = cellexalObj@index
   save( facs, file= "facs.RData")
   rm( facs )
   
   cell.ids = cellexalObj@meta.cell
   save( cell.ids, file="cell.ids.RData")
   rm(cell.ids)
   
   diff.proj = cellexalObj@drc$diffusion
   save( diff.proj, file="diff.proj.RData")
   rm( diff.proj )
   
   ddr.proj = cellexalObj@drc$DDRtree
   save( ddr.proj, file="ddr.proj.RData" )
   rm(ddr.proj)
   
   tsne.proj = cellexalObj@drc$tSNE
   save( tsne.proj, file="tsne.proj.RData")
   rm(tsne.proj)
   
}
```

Then load the data:
```{r, fig.show='hold',eval=TRUE}
	
load("log2data.RData")
load("facs.RData")
load("cell.ids.RData")
load("diff.proj.RData")
load("ddr.proj.RData")
load("tsne.proj.RData")
```

`log2data` is a matrix of highly variable genes in log2 form. The first 10 columns and 10 rows look as so:

```{r, fig.show='hold',eval=T}
log2data[1:10,1:10]
```
Note the cell IDs are in the column names, and the gene names are the row names.

`facs` is a matrix of cell surface marker intensities captured during index sorting:

```{r, fig.show='hold',eval=T}
head(facs)
```
Cell IDs are in the row names, and the name of the surface protein in the columns.

`cell.ids` is a 1/0 matrix that assigns metadata to the each cell. In this case it shows the cell type 
each cell was sorted as:

```{r, fig.show='hold',eval=T}
head(cell.ids)
```
Cell IDs are in the row names, and the name of the meta data in the columns. A 1 is given if the cell 
belongs to a metadata class, 0 otherwise.

`diff.proj`,`tsne.proj`,and `diff.proj` are the results from three different dimension reduction methods 
applied to `log2data`, specifically DDRTree, tSNE and diffusion map respectively. Each is a three column matrix 
of x/y/z coordinates. For example:

```{r, fig.show='hold',eval=T}
head(diff.proj)
```

The first step is to put all the dimension reduction outputs into a single list:

```{r, fig.show='hold',eval=T}
proj.list <- list(diffusion=diff.proj,DDRtree=ddr.proj,tSNE=tsne.proj)
```

The next is to make a `cellexalvr` object by calling `MakeCellexalVRObj` and passing the required objects to it:

```{r, fig.show='hold',eval=T}
cellvr <- MakeCellexalVRObj( Matrix::Matrix(log2data, sparse=TRUE),
  drc.list=proj.list,specie="mouse",cell.meta=cell.ids,facs.data=NULL)
```
In the same step we also set the specie as mouse which ensures the correct 
transcription factor IDs are using during network construction if implemented in CellexalVR.

Calling the object name will detail it's contents:

```{r}
cellvr
```

The last step is to call `export2cellexalvr` which will write the neccessary files from the `cellvr` object 
to a specified directory:

```{r, fig.show='hold',eval=F}
export2cellexalvr(cellvr,"CellexalOut/")
```
Make sure that this directory exists.

All the files needed by CellexalVR are created at this point. The entire "CellexalOut/" folder should then be 
moved/copied to the "Data" folder in your CellexalVR setup. See the [manual](https://www.cellexalvr.med.lu.se/manual)
 for more details including a schematic of the folder structure.

## Making a `cellexalvr` object from scratch
While `MakeCellexalVRObj` is the most convenient way to make the object, sometimes you want to make one manually. 
This is done calling `new`:

```{r,fig.show='hold',eval=T}
cell.vr.scr <- new("cellexalvr",data=Matrix::Matrix(log2data,sparse=T),drc=list(tsne=tsne.proj))
cell.vr.scr
```

We can add another set of dimension reduction coordinates using the `addMDS2cellexalvr` function:

```{r,fig.show='hold',eval=T}
cell.vr.scr <- addDRC2cellexalvr(cell.vr.scr,ddr.proj,"DDRTree")
cell.vr.scr
```

To add metadata for each cell use `addCellMeta2cellexalvr`:
```{r,fig.show='hold',eval=T}
cell.vr.scr <- addCellMeta2cellexalvr(cell.vr.scr,cell.ids)
cell.vr.scr

```

To add FACS for each cell use `addFACS2cellexalvr`:
```{r,fig.show='hold',eval=F}
cell.vr.scr <- addFACS2cellexalvr(cell.vr.scr,facs)
```

Setting the specie is done using the `set.specie` function:

```{r,fig.show='hold',eval=T}
cell.vr.scr <- set.specie(cell.vr.scr,"mouse")
cell.vr.scr
```

##Making cell metadata from a data frame

CellexalVR requires metadata in the form of a 1/0 matrix, but many packages store it as a data frame. 
CellexalvrR has function to convert a data frame into a 1/0 matrix. First, lets make a data frame:

```{r,fig.show='hold',eval=T}
meta.df <- data.frame(CellType=sample(c("Type1","Type2","Type3"),10,replace=T),
                      Phase=sample(c("G1","G2M","S"),10,replace=T),
                      Treatment=sample(c("WT","Dox"),10,replace=T))
head(meta.df)

```

We can now make a correctly formatted cell metadata matrix by applying the function `make.cell.meta.from.df` 
using only the fields we need, in this case the "CellType" and "Treatment" columns:
```{r,fig.show='hold',eval=T}

required.cell.metad <- make.cell.meta.from.df(meta.df,c("CellType","Treatment"))
head(required.cell.metad)

```

It can be seen the field name is placed in the column heading preceeding a "@", and this is used by 
CellexalVR to form catagories on the menu system, so "CellType" and "Treatment" will be on separate tabs. 
This metadata matrix can now be added to a `cellexalvrR` object as decribed above using the `addCellMeta2cellexalvr` 
function.

##Converting a Seurat object to a cellexalvr object

For those using [Seurat](https://satijalab.org/seurat/), cellexalvrR offers some basic functionality 
to convert a Seurat object to a cellexalvr object. It takes the data from the `@dat` slot, and cell 
identities from the `@ident` slot, and will also take the cell-cycle phase assignments if this has been done.
 If PCA has been performed the first three components are taken as MDS coordinates, and the same if tSNE has
  been implemented using the `dim.embed=3` option when calling `RunTSNE` from Seurat.

```{r,eval=F}
# mca comes from from https://satijalab.org/seurat/v3.0/mca.html
mca <- RunUMAP(mca, dims = 1:75, min.dist = 0.75,n.components=3)

mca.data <- GetAssayData(object = mca) #Extract the expression data
drl <- list(UMAP=Embeddings(object = mca, reduction = "umap")) # Put the UMAP coordinated into a list
meta <- make.cell.meta.from.df(mca[[]],"Tissue") # Make the metadata using just the "Tissue" column

cvr <- new("cellexalvrR",data=mca.data,drc=drl) # Initialise a new cellexalvrR object with the expression data and UMAP
cvr <- set.specie(cvr,"mouse") # Set the specie to Mouse
cvr <- addCellMeta2cellexalvr(cvr,meta) # Add the metadata to the cellexalvrR object
export2cellexalvr(cvr,"MCA_full") #Export the files to a folder called "MCA_full" 

```






